test/dns-test.js

```diff
@@ -0,0 +1,396 @@
+      assert.equal(answer.name, name);
+    });
+  });
+
+  describe('Authoritative Resolver', () => {
+    it('should return no answer for an ICANN rooted name', async () => {
+      const name = 'com.';
+
+      const res = await nstub.lookup(name, types.A);
+
+      assert.equal(res.question[0].name, name);
+      assert.equal(res.answer.length, 0);
+
+      // Handshake zone is not authoritative for ICANN based name
+      assert.equal(res.aa, false);
+    });
+
+    it('should return no answer for HNS name pre-update', async () => {
+      const name = 'foo.';
+
+      const res = await nstub.lookup(name, types.A);
+      assert.equal(res.answer.length, 0);
+    });
+
+    // record to save in the authenticated tree
+    for (const [hstype, item] of Object.entries(json)) {
+      const type = types[item.type];
+      const resource = Resource.fromJSON(item.resource);
+
+      // args: size, height, network
+      const name = rules.grindName(5, 1, network);
+
+      it(`should update ${name}`, async () => {
+        await wallet.client.post(`/wallet/${wallet.id}/open`, {
+          name: name
+        });
+
+        await mineBlocks(treeInterval + 1, coinbase);
+
+        await wallet.client.post(`/wallet/${wallet.id}/bid`, {
+          name: name,
+          bid: 1000,
+          lockup: 2000
+        });
+
+        await mineBlocks(biddingPeriod + 1, coinbase);
+
+        await wallet.client.post(`/wallet/${wallet.id}/reveal`, {
+          name: name
+        });
+
+        await mineBlocks(revealPeriod + 1, coinbase);
+
+        await wallet.client.post(`/wallet/${wallet.id}/update`, {
+          name: name,
+          data: resource.toJSON()
+        });
+
+        // mine a block
+        await mineBlocks(1, coinbase);
+
+        const rinfo = await nclient.execute('getnameresource', [name]);
+
+        const json = resource.getJSON(name);
+        assert.deepEqual(rinfo, json);
+
+        const info = await nclient.execute('getnameinfo', [name]);
+        const data = Buffer.from(info.info.data, 'hex');
+
+        const returned = Resource.decode(data);
+        assert.deepEqual(returned.getJSON(name), json);
+
+        // this commits the tree state to disk
+        // updates served by the dns server
+        await mineBlocks(treeInterval, coinbase);
+      });
+
+      it(`should return authenticated ${hstype} record`, async () => {
+        // Certain types of queries require the
+        // name to be formatted with additional data
+        const query = buildQuery(name, resource, type);
+
+        const response = await nstub.lookup(query, type);
+        const res = response.toJSON();
+        const dns = resource.toDNS(util.fqdn(name), type).toJSON();
+
+        // query for the zone signing key
+        const dnskey = await nstub.lookup('.', types.DNSKEY);
+
+        // parse the zone signing key and
+        // key signing key out of the response
+        const {zsk, ksk} = getSigningKeys(dnskey);
```

Comment:
Do these need to run on each loop? Or can they be hoisted

Reply:
We know that technically the `DNSKEY` record shouldn't change, sending a request for it each time is extra paranoia. In practice the `DNSKEY` could be rolled over for a zone. It takes awhile for the root zone key to roll over, I wonder how often other zone keys roll over.